l1-contracts/contracts/bridge/L1ERC20Bridge.sol

```diff
@@ -60,73 +76,253 @@ contract L1ERC20Bridge is IL1Bridge, IL1BridgeLegacy, ReentrancyGuard {
     /// @param _deployBridgeImplementationFee How much of the sent value should be allocated to deploying the L2 bridge
     /// implementation
     /// @param _deployBridgeProxyFee How much of the sent value should be allocated to deploying the L2 bridge proxy
-    function initialize(
+    function startErc20BridgeInitOnChain(
+        uint256 _chainId,
         bytes[] calldata _factoryDeps,
-        address _l2TokenBeacon,
-        address _governor,
         uint256 _deployBridgeImplementationFee,
         uint256 _deployBridgeProxyFee
-    ) external payable reentrancyGuardInitializer {
-        require(_l2TokenBeacon != address(0), "nf");
-        require(_governor != address(0), "nh");
-        // We are expecting to see the exact three bytecodes that are needed to initialize the bridge
-        require(_factoryDeps.length == 3, "mk");
-        // The caller miscalculated deploy transactions fees
-        require(msg.value == _deployBridgeImplementationFee + _deployBridgeProxyFee, "fee");
-        l2TokenProxyBytecodeHash = L2ContractHelper.hashL2Bytecode(_factoryDeps[2]);
-        l2TokenBeacon = _l2TokenBeacon;
-
+    ) external payable {
+        {
+            require(l2BridgeAddress[_chainId] == address(0), "L1EB B deployed");
+            // We are expecting to see the exact three bytecodes that are needed to initialize the bridge
+            require(_factoryDeps.length == NUMBER_OF_FACTORY_DEPS, "L1EB w # of f-deps");
+            require(factoryDepsHash == keccak256(abi.encode(_factoryDeps)), "L1EB w f-deps");
+            require(bridgeProxyDeployOnL2TxHash[_chainId] == 0x00, "L1EB b. proxy tx sent"); // clear the tx first by proving the txs succeeded or failed
+        }
+        bool ethIsBaseToken = bridgehub.baseToken(_chainId) == ETH_TOKEN_ADDRESS;
+        {
+            if (!ethIsBaseToken) {
+                require(msg.value == 0, "L1EB m.v > 0, e base");
+            }
+        }
         bytes32 l2BridgeImplementationBytecodeHash = L2ContractHelper.hashL2Bytecode(_factoryDeps[0]);
         bytes32 l2BridgeProxyBytecodeHash = L2ContractHelper.hashL2Bytecode(_factoryDeps[1]);
+        {
+            // Deploy L2 bridge implementation contract
+            (address bridgeImplementationAddr, bytes32 bridgeImplTxHash) = BridgeInitializationHelper
+                .requestDeployTransaction(
+                    ethIsBaseToken,
+                    _chainId,
+                    bridgehub,
+                    _deployBridgeImplementationFee,
+                    l2BridgeImplementationBytecodeHash,
+                    "", // Empty constructor data
+                    _factoryDeps // All factory deps are needed for L2 bridge
+                );
+
+            // Prepare the proxy constructor data
+            bytes memory l2BridgeProxyConstructorData;
+            {
+                address owner = owner();
+                address l2Owner = AddressAliasHelper.applyL1ToL2Alias(owner);
```

Comment:
If the owner is an EOA, then access is fully lost, even if usually it is contract, better to double check.

I also don't like that L1 owner is not in sync with L2 owners. If L1 owner changed L2 owners still can perform L2 upgrade. With having many hyperchains it is better to have proxy for the owner (not in term of delegate calls) but contract that is the owner for all L2 contracts and L1 owner is the owner of that proxy.

Reply:
Ok, I will add the same check as above. 

I agree with the second point, for that we don't have to change anything right?

Reply:
> I agree with the second point, for that we don't have to change anything right?

Because Governance is the owner of L1Erc20Bridge?

Reply:
I mean change anything here. We can deploy the proxy Owner and set it as governor on here and on the L2 as needed.